lib/post_creator.rb

```diff
@@ -111,38 +111,13 @@ def valid?
         return false
       end
 
-      # Make sure none of the users have muted or ignored the creator
-      users = User.where(username_lower: names).pluck(:id, :username).to_h
-
-      User
-        .joins("LEFT JOIN user_options ON user_options.user_id = users.id")
-        .joins("LEFT JOIN muted_users ON muted_users.user_id = users.id AND muted_users.muted_user_id = #{@user.id.to_i}")
-        .joins("LEFT JOIN ignored_users ON ignored_users.user_id = users.id AND ignored_users.ignored_user_id = #{@user.id.to_i}")
-        .where("user_options.user_id IS NOT NULL")
-        .where("
-          (user_options.user_id IN (:user_ids) AND NOT user_options.allow_private_messages) OR
-          muted_users.user_id IN (:user_ids) OR
-          ignored_users.user_id IN (:user_ids)
-        ", user_ids: users.keys)
-        .pluck(:id).each do |m|
-
-        errors.add(:base, I18n.t(:not_accepting_pms, username: users[m]))
-      end
-
-      # Is Allowed PM users list enabled for any recipients?
-      users_with_allowed_pms = allowed_pms_enabled(users).pluck(:id).uniq
-
-      # If any of the users has allowed_pm_users enabled check to see if the creator
-      # is in their list
-      if users_with_allowed_pms.any?
-        users_sender_can_pm = allowed_pms_enabled(users)
-          .where("allowed_pm_users.allowed_pm_user_id" => @user.id.to_i)
-          .pluck(:id).uniq
+      # Make sure none of the users have muted or ignored the creator or prevented
+      # PMs from being sent to them
+      target_user_communication_preferences = UserCommunicationDefender.new(acting_user_id: @user.id, target_usernames: names).fetch_user_preferences
```

Comment:
Design thought: Is the wording/naming too lengthy?

Reply:
Maybe an API like this? `user_comm_screener.screen(usernames: names)` which only returns the `names` that are valid.

Reply:
Oh yeah I see, so only ones that can communicate in _any_ way as a shorthand. The only downside to returning the valid ones is that then you'd have to compare that with the original array to get the invalid ones. Instead something like this would be better I think:

```ruby
UserCommScreener.new(acting_user_id: @user.id, target_usernames: names).each_preventing_comms do |username|
  errors.add(:base, I18n.t(:not_accepting_pms, username: username))
end
```

Reply:
Then we can have other ones like `.each_ignoring_or_muting` , and have an array version of it too which is just `.ignoring_or_muting`. Probably would be even easier to just have the array versions and rely on the caller to do `.each` on the end

Reply:
Yup I think returning an array would be better and let the caller deal with it. It'll be more flexible.